opencog/attentionbank/AttentionBank.cc

```diff
@@ -320,21 +321,23 @@ void AttentionBank::updateAttentionalFocus(const Handle& h,
 
 Handle AttentionBank::getRandomAtomNotInAF(void)
 {
-    std::lock_guard<std::mutex> lock(AFMutex);
-    auto find_in_af = [this](const Handle& h){
-       auto it = std::find_if(attentionalFocus.begin(), attentionalFocus.end(),
-                [h](const std::pair<Handle, AttentionValuePtr>& hsp) { return hsp.first == h; });
-       return it;
-    };
-
-    int count = 50; // We might get stuck in the while loop if there are no atoms outside the AF
-    Handle h = Handle::UNDEFINED;
-    do {
-        h = _importanceIndex.getRandomAtom();
-    } while(find_in_af(h) != attentionalFocus.end() and --count > 0);
-    // Make sure the last selection did not select from the AF.
-    if(find_in_af(h) != attentionalFocus.end())
+    UnorderedHandleSet atoms_not_in_af = _importanceIndex.getHandleSet(0);
+
+    {
+        std::lock_guard<std::mutex> lock(AFMutex);
```

Comment:
I think this lock should be scoped global to this function since an atom which was not in the attentionalFocus might be inserted as soon as the lock is released while the remaining code assumes it is not in the attentionalFocus and selects it.

Reply:
Making scope of the lock global will not make a big difference as atom can be inserted after leaving critical section but before returning to the caller function.
I would say that presence of two independent locks _mtx and AFMutex is more serious problem. _importanceIndex and attentionalFocus can be independently updated by different threads, not all _importanceIndex updates are guarded by lock, but I would say it is separate issue.

Reply:
the _mtx lock is used for synchronizing the funds. Maybe it would be good if you explain the situation in the issue you opened. Thanks.

Reply:
added comment to issue https://github.com/opencog/atomspace/issues/1658